### PR TITLE
Propagate platform to all marshals.

### PR DESCRIPTION
**Pull request by sipsma**
Signed-off-by: Erik Sipsma <erik@sipsma.dev>

Should fix the problems when running from macos.

Opened an issue for automating that: https://github.com/dagger/dagger/issues/3225

But for now verified by running directly from macos. 

This will also be a little bit of setup for integration the multiplatform PR w/ the new API, so no throwaway here hopefully.

@vito the rebase w/ your PRs will be painful sadly. I had one important comment on your PR ([here](https://github.com/dagger/dagger/pull/3205#discussion_r986333125)), but let's try to merge that first then I'll do the rebase on this one if possible.

### CODE REVIEW

**Comment by sipsma** [comment]
Also @vito the linter scolded me for having too many return args from `Container.Decode()`, specifically when I had to ignore most of them in some calls ([this linter called dogsled](https://github.com/alexkohler/dogsled)).

So the second commit just has `Container.Decode` return the whole payload object. This slightly deviates it from the pattern established elsewhere, let me know if you think that's worth re-figuring out.

File: core/container.go [hunk 1]
```diff
@@ -153,12 +157,12 @@ func (container *Container) WithMountedDirectory(ctx context.Context, target str
 		return nil, err
 	}
 
-	dirSt, dirRel, err := source.Decode()
+	dirSt, dirRel, dirPlatform, err := source.Decode()
```
Comment (vito) [inline core/container.go:160]: Good thing we had the return value budget for this. :joy:

File: core/container.go [hunk 2]
```diff
@@ -452,7 +457,7 @@ type ContainerExecOpts struct {
 }
 
 func (s *containerSchema) exec(ctx *router.Context, parent *Container, args containerExecArgs) (*Container, error) {
-	return parent.Exec(ctx, s.gw, s.platform, args.Args, args.Opts)
```
Comment (vito) [inline core/container.go:455]: I like that this went away, it felt a little odd that this was the only thing that needed it. (Turns out...)

**Review by vito — APPROVED** [review]
Thanks for fixing this! The changes look great.

Looking forward to multi-platform support being merged, I'm guessing its tests would have caught this. (It's also just, like, cool.)

**Comment by sipsma** [comment]
> Looking forward to multi-platform support being merged, I'm guessing its tests would have caught this. (It's also just, like, cool.)

Yeah we gotta settle on which approach to take ([discussion](https://github.com/dagger/dagger/pull/3119#discussion_r984028111)) and at this point it's probably worth just rebasing it entirely on the new API since the old one's expiration date is approaching fast. Fortunately this PR is a little bit of a layup for all that.

**Pull request merged**